include/wil/filesystem.h

```diff
@@ -1043,8 +1046,224 @@ namespace wil
         THROW_IF_FAILED(GetFileInfoNoThrow<infoClass>(fileHandle, result));
         return result;
     }
+
+    // Helpers to make the CreateFileW API easier to use. This segregates the OPEN_EXISTING "open an existing file"
+    // cases from the "create a new file" that has 4 variations represented in the create_file_create_behavior enum.
+
+    // https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew?devlangs=cpp&f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(FILEAPI%252FCreateFileW)%3Bk(CreateFileW)%3Bk(DevLang-C%252B%252B)%3Bk(TargetOS-Windows)%26rd%3Dtrue
+
+    struct file_and_error_result
+    {
+        file_and_error_result(HANDLE file_handle, DWORD error) : file(file_handle), last_error(error)
+        {
+        }
+
+        wil::unique_hfile file;
+        DWORD last_error{};
+    };
+
+    /** Non-throwing open existing using OPEN_EXISTING.
+    ~~~
+    auto handle = wil::try_open_file(filePath.c_str());
+    ~~~
+    */
+    inline file_and_error_result try_open_file(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+        DWORD dwShareMode = FILE_SHARE_READ,
+        bool inheritHandle = false,
+        DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL) noexcept
+    {
+        SECURITY_ATTRIBUTES secAttributes{ sizeof(secAttributes) };
+        secAttributes.bInheritHandle = inheritHandle;
+        return { CreateFileW(path, dwDesiredAccess, dwShareMode, &secAttributes, OPEN_EXISTING, dwFlagsAndAttributes, nullptr),
+            ::GetLastError() };

```

Comment:
It might be worth conditionally calling `GetLastError` depending on whether or not `CreateFileW` returned an invalid handle (& deduce `ERROR_SUCCESS` if not).

Reply:
That's a good point.  Won't the value of GetLastError be undefined on success?  Or is `CreateFileW` guaranteed to clear it to success on entry?

Reply:
From the documentation on [GetLastError](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror#remarks):

> some functions call SetLastError with a zero when they succeed, wiping out the error code set by the most recently failed function.

Sounds like it's not guaranteed. The only relevant info on the documentation for [CreateFileW](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew#return-value) seems to be:

> If the function fails, the return value is INVALID_HANDLE_VALUE. To get extended error information, call [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror).

There's nothing that indicates that the function calls `SetLastError` for success cases

Reply:
the docs are clear that the last error is meaningful in success cases and that ensures it always is.

Reply:
Ahh, I missed a few mentions of when error values are set:
> If the specified file exists and is writable, the function truncates the file, the function succeeds, and last-error code is set to ERROR_ALREADY_EXISTS

I'm still of the opinion that most of the time this information is not needed, but it makes more sense now. Alternatively, and this might be over-designing things, we could only return the last set error for the functions where it makes sense (`CREATE_ALWAYS` and `OPEN_ALWAYS`). We could also possibly "translate" this into something like a `bool already_exists` instead of an error code.